Strange error reported when using `facet(derive(Default))`

Getting an odd set of errors when trying to derive facet-default. Repro comes after the errors.

```
❯ cargo check
    Checking evil-repro v0.1.0 (/tmp/evil-repro)
error[E0107]: missing generics for struct `IdType`
  --> src/main.rs:20:12
   |
20 | pub struct IdType<TKind: private::Sealed> {
   |            ^^^^^^ expected 1 generic argument
   |
note: struct defined here, with 1 generic parameter: `TKind`
  --> src/main.rs:20:12
   |
20 | pub struct IdType<TKind: private::Sealed> {
   |            ^^^^^^ -----
help: add missing generic argument
   |
20 | pub struct IdType<TKind><TKind: private::Sealed> {
   |                  +++++++

For more information about this error, try `rustc --explain E0107`.
error: could not compile `evil-repro` (bin "evil-repro") due to 1 previous error
```

```rs
use facet::Facet;
use std::marker::PhantomData;

mod private {
    use facet::Facet;

    pub(super) trait Sealed {
        type Associated: for<'facet> Facet<'facet>;
    }
}

#[derive(Facet)]
pub struct IdMarker;
impl private::Sealed for IdMarker {
    type Associated = String;
}

#[derive(Facet)]
#[facet(derive(Default))]
pub struct IdType<TKind: private::Sealed> {
    something: String,
    _kind: PhantomData<TKind>,
}
```

Commenting out the `#[facet(derive(Default))]` line...
```
❯ cargo check
    Checking evil-repro v0.1.0 (/tmp/evil-repro)
warning: trait `Sealed` is more private than the item `IdType`
  --> src/main.rs:20:1
   |
20 | pub struct IdType<TKind: private::Sealed> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ struct `IdType` is reachable at visibility `pub`
   |
note: but trait `Sealed` is only usable at visibility `pub(crate)`
  --> src/main.rs:7:5
   |
 7 |     pub(super) trait Sealed {
   |     ^^^^^^^^^^^^^^^^^^^^^^^
   = note: `#[warn(private_bounds)]` on by default

warning: `evil-repro` (bin "evil-repro") generated 1 warning
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.21s
```
it checks just fine. Sure, there's that private_bounds thing, but that's fine.

Can provide more information upon request; unsure what to add. Latest (as of issue opening) versions of `facet` and `facet-default` were used.

--- a/facet-default/src/lib.rs
+++ b/facet-default/src/lib.rs
@@ -86,7 +86,7 @@ macro_rules! __facet_invoke {
                 @plugin {
                     @name { "Default" }
                     @template {
-                        impl ::core::default::Default for @Self {
+                        impl @GenericParams ::core::default::Default for @Self @WhereClause {
                             fn default() -> Self {
                                 @if_struct {
                                     Self {

--- a/facet-default/tests/integration.rs
+++ b/facet-default/tests/integration.rs
@@ -2,6 +2,7 @@
 
 use facet::Facet;
 use facet_default as default;
+use std::marker::PhantomData;
 
 /// Test struct with various default attributes
 #[test]
@@ -225,4 +226,33 @@ fn test_string_literal_default() {
 
     let blah = Blah::default();
     assert_eq!(blah.field, "ha");
+}
+
+/// Repro for issue #2398: `@Self` in plugin templates must include generic
+/// arguments without their bounds, e.g. `IdType<TKind>` rather than `IdType`.
+#[test]
+fn test_derive_default_with_bounded_generic() {
+    mod private {
+        use facet::Facet;
+
+        pub(super) trait Sealed {
+            type Associated: for<'facet> Facet<'facet>;
+        }
+    }
+
+    #[derive(Facet)]
+    pub struct IdMarker;
+    impl private::Sealed for IdMarker {
+        type Associated = String;
+    }
+
+    #[derive(Facet)]
+    #[facet(derive(Default))]
+    pub struct IdType<TKind: private::Sealed> {
+        something: String,
+        _kind: PhantomData<TKind>,
+    }
+
+    let id = IdType::<IdMarker>::default();
+    assert_eq!(id.something, String::default());
 }

--- a/facet-macros-impl/src/plugin.rs
+++ b/facet-macros-impl/src/plugin.rs
@@ -815,6 +815,8 @@ fn handle_directive(
     match directive.as_str() {
         // === Type-level directives ===
         "Self" => emit_self_type(ctx, output),
+        "GenericParams" => emit_generic_params(ctx, output),
+        "WhereClause" => emit_where_clause(ctx, output),
 
         // === Looping directives ===
         "for_variant" => handle_for_variant(iter, ctx, output),
@@ -862,7 +864,25 @@ fn handle_directive(
 
 fn emit_self_type(ctx: &EvalContext<'_>, output: &mut TokenStream) {
     let name = ctx.parsed_type.name();
-    output.extend(quote! { #name });
+    let generics = match ctx.parsed_type {
+        facet_macro_parse::PType::Struct(s) => s.container.bgp.display_without_bounds(),
+        facet_macro_parse::PType::Enum(e) => e.container.bgp.display_without_bounds(),
+    };
+    output.extend(quote! { #name #generics });
+}
+
+fn emit_generic_params(ctx: &EvalContext<'_>, output: &mut TokenStream) {
+    let generics = match ctx.parsed_type {
+        facet_macro_parse::PType::Struct(s) => s.container.bgp.display_with_bounds(),
+        facet_macro_parse::PType::Enum(e) => e.container.bgp.display_with_bounds(),
+    };
+    output.extend(quote! { #generics });
+}
+
+fn emit_where_clause(_ctx: &EvalContext<'_>, _output: &mut TokenStream) {
+    // TODO: facet-macro-parse does not currently expose user where-clauses in
+    // PContainer. This directive is reserved so plugin templates can place a
+    // where-clause once parsing support is available.
 }
 
 // ============================================================================

